include/openthread/dns.h

```diff
@@ -140,6 +141,93 @@ void otDnsInitTxtEntryIterator(otDnsTxtEntryIterator *aIterator, const uint8_t *
  */
 otError otDnsGetNextTxtEntry(otDnsTxtEntryIterator *aIterator, otDnsTxtEntry *aEntry);
 
+/**
+ * This function is called when a DNS-SD query subscribes a service or service instance.
+ *
+ * @param[in] aContext      A pointer to the application-specific context.
+ * @param[in] aFullName     The null-terminated full service name (e.g. "_ipps._tcp.default.service.arpa."), or
+ *                          full service instance name (e.g. "OpenThread._ipps._tcp.default.service.arpa.").
```

Comment:
How would we represent the name if the `instance name` label has a `.` in it (which is allowed)?
We can consider providing the instance label as a separate string to make it easier for the user.

Thought?

Reply:
In the current interface design, the implementation is responsible for identifying what `@p aFullName` is. It can be either of:
1. service name
2. service instance name
3. host name

I think it's possible to correctly identify the `instance name` even when it contains a `.` in it. So, the interface should have included enough information. 

The benefit of the current interface is that the DNS-SD server in OT core don't need to further split the full name into `instance name` and `full service name`, which needs 256B buffers. 

The alternative is to provide different interfaces for different names:
1. otDnssdQuerySubscribeService(aFullServiceName)
2. otDnssdQuerySubscribeServiceInstance(aFullServiceName, aInstanceName)
3. otDnssdQuerySubscribeHost(aFullHostName)

Together with the 3 corresponding "Unsubscribe" APIs, there will be totally 6 callbacks to setup. I prefer the current interfaces (with only 2 callbacks) because it's simpler for the OT core. 

Thoughts?

Reply:
The current model SGTM.